src/vector3d.cpp

```diff
@@ -0,0 +1,288 @@
+ * @return                                     Vector in previous basis
+ */
+Vector3d Vector3d::RotateVector(const double p_ThetaE, const double p_PhiE, const double p_PsiE) {
+
+    // Replace for convenience, undefine below
+    #define cTheta cos(p_ThetaE)
+    #define cPhi   cos(p_PhiE)
+    #define cPsi   cos(p_PsiE)
+    #define sTheta sin(p_ThetaE)
+    #define sPhi   sin(p_PhiE)
+    #define sPsi   sin(p_PsiE)
+
+    // Define the Rotation Matrix     
+    std::vector<DBL_VECTOR> RotationMatrix = {
+        { cPhi*cPsi - sPhi*cTheta*sPsi ,  -cPhi*sPsi - sPhi*cTheta*cPsi ,  sTheta*sPhi },
+        { sPhi*cPsi + cPhi*cTheta*sPsi ,  -sPhi*sPsi + cPhi*cTheta*cPsi , -sTheta*cPhi },
+        { sTheta*sPsi                  ,  sTheta*cPsi                   ,  cTheta      }};
+
+    #undef cTheta
+    #undef cPhi
+    #undef cPsi
+    #undef sTheta
+    #undef sPhi
+    #undef sPsi
+
+
+    // Multiply RotationMatrix * p_oldVector
+    Vector3d oldVector = (*this);
+    Vector3d newVector = Vector3d(0.0, 0.0, 0.0);
+
+    for (int i=0; i< 3; i++) {
+        for (int j=0; j<3; j++) {
+            newVector[i] += RotationMatrix[i][j] * oldVector[j];
+        }
+    }
+
+    return newVector;
+}
+
+
+
+
+
+///////////////////////////////////////////////////////////
+// 
+// Linear Algebra Functions
+// 
+///////////////////////////////////////////////////////////
+
+namespace linalg {
+
+    /*
+     * Calculate the standard dot product of two vectors
+     *
+     *
+     * @param   [IN]   a                            first vector
+     * @param   [IN]   b                            second vector
+     * @return                                      dot product
+     */
+    double dot(const Vector3d& a, const Vector3d& b) {
+    
+        double c = 0;
+    
+        for (int i=0; i<3; i++) {
+            c += a[i]*b[i];
+        }
+    
+        return c;
+    }
+    
+    
+    /*
+     * Calculate the standard cross product of two vectors
+     *
+     *
+     * @param   [IN]   a                            first vector
+     * @param   [IN]   b                            second vector
+     * @return                                      cross product
+     */
+    Vector3d cross(const Vector3d& a, const Vector3d& b) {
+    
+        Vector3d c = Vector3d(0, 0, 0);
+    
+        c[0] = a[1]*b[2] - a[2]*b[1];
+        c[1] = a[2]*b[0] - a[0]*b[2];
+        c[2] = a[0]*b[1] - a[1]*b[0];
+    
+        return c;
+    }
+
+    /*
+     * Calculate the angle between two vectors.
+     *
+     *
+     * @param   [IN]   a                            first vector
+     * @param   [IN]   b                            second vector
+     * @return                                      angle between them, in radians
+     */
+    double angleBetween(const Vector3d& a, const Vector3d& b) {
+        // Angle between 2 vectors, between [0, 2pi] 
+        return acos(linalg::dot(a,b));
+    }
+
+}
+
```

Comment:
Did you do sanity checks for these linear algebra functions? If so, good. If not, please do.

Reply:
Yes, all checks passed.